File: src/08.dbpointer/34.287.ts
export function findDuplicate(nums: number[]): number {
    let slow = 0;
    let fast = 0;

    slow = nums[slow];
    fast = nums[nums[fast]];

    while (fast != slow) {
        slow = nums[slow];
        fast = nums[nums[fast]];
    }

    // 寻找环的入口
    let pre1 = 0;
    let pre2 = slow;
    while (pre1 != pre2) {
        pre1 = nums[pre1];
        pre2 = nums[pre2];
    }

    return pre1;
}

File: src/08.dbpointer/34.287.go
package dbpointer

func FindDuplicate(nums []int) int {
	fast, slow := 0, 0
	slow = nums[slow]
	fast = nums[nums[fast]]

	for slow != fast {
		slow = nums[slow]
		fast = nums[nums[fast]]
	}

	pre := 0
	for pre != slow {
		pre = nums[pre]
		slow = nums[slow]
	}

	return pre
}
